Add screen shot all devices functionality

durgakiran opened the pull request:
# ✨ Pull Request

### 📓 Referenced Issue
closes #541 

### ℹ️ About the PR
Give take a screenshot of all device Web views at once, including merging of images.

Features:
<h2>

```diff
+ Choosing screens to capture
```
</h2>

![image](https://user-images.githubusercontent.com/17452039/117538243-e8709100-b022-11eb-9568-547fa35799f1.png)

<h2>

```diff
+ Capture fail and recovery mechanism
```
</h2>

One of the existing issues with the existing screenshot mechanism is it extremely fail prone. The main reason as I observed is crash electron main render thread, which fails and never recovers when continuously `scrollViewPort` function is called. One more issue with the current screenshot capture is if the render thread fails to recover entire application needs to be reloaded.

To counter such fails we introduce a function that returns a promise with a retry mechanism limited to 3 retries. The function listens to webContents `crashed` event. we use `crashed` event to understand the if webview is killed or just crashed. if Webview is just crashed, the code will again load the crashed web view. Once webview reloads function will listen to `dom-ready` event. which notifies once the web view is completely loaded, then we try to capture the image again.

```
captureWithRetry(retry: number = 3): Promise<Electron.NativeImage> {
    return new Promise((resolve, reject) => {
      let retryCount = 0;
      const address = this.webView.getAttribute('src');

      const handler = (event, killed) => {
        if (!killed) {
          this.webView.setAttribute('src', address);
        } else {
          reject(new Error('WebView killed.... : ('));
        }
      };
      this.webContents.on('crashed', handler);

      const callGetImage = () => {
        // noinspection JSIgnoredPromiseFromCall
        getImage();
      };

      this.webContents.on('dom-ready', callGetImage);

      const getImage = async () => {
        if (retryCount === retry) {
          reject(new Error('Unable to produce image'));
        }
        retryCount += 1;
        const image = await this.captureFullHeightScreenShot();
        this.webContents.removeListener('dom-ready', callGetImage);
        this.webContents.removeListener('crashed', handler);
        resolve(image);
      };

      // noinspection JSIgnoredPromiseFromCall
      getImage();
    });
  }
```
If Image capture fails beyond the threshold i.e, 3. The code will skip that particular web view and proceed with other web views.
This way we are not abandoning the app in a failed state.

<h2>

```diff
+ Merging images
```
</h2>
Each web view, whose image is captured, binary data of the image is store in the memory. Once completion of taking individual screenshots is done. The binary data of each data is combined by drawing on a canvas. The canvas data again converted to binary and saved to the folder. The merging of images **does not** block Electron's main threads.

<h2>

```diff
! Caution: Merging of images takes a lot of time
```
</h2>

<h2>

```diff
+ Final indication
```
</h2>
Once the capture and saving process is over, automatically the created folder is opened.

**There are some improvements in code refactoring and structure, I think those are self-evident in the PR file changes.**

### 🖼️ Testing Scenarios / Screenshots
[please click here to view feature demo video](https://www.loom.com/share/70cb2f3df14d474683d2f4170c5708e2).

Example merged screenshots.
![Google-all-2021-10-5](https://user-images.githubusercontent.com/17452039/117601052-e580b800-b16a-11eb-8056-04f7394e0d00.png)

ziabs committed:
Git ignore .vscode

Git ignore update for .vscode

rishichawda committed:
feat(desktop-app/app): show url link on hover

The url from href property for the anchor element is displayed in the bottom left of the screen on
hover

fix #516

rishichawda committed:
improvement(components/linkhoverdisplay): display full url, smoother animation

rishichawda committed:
fix(linkhoverdisplay): fix url text color on light theme

karllabrador committed:
Added disable spellcheck userpref, but needs a restart

karllabrador committed:
Added restart required note on toggle spellcheck

manojVivek committed:
Merge pull request #564 from responsively-org/all-contributors/add-jcabak

docs: add jcabak as a contributor

Commit:
Fixing the vulnerability where local files contents can be accessed by the remote page

durgakiran committed:
Add event disable/enable event mirroring

Commit:
Merge remote-tracking branch 'origin/master' into master

manojVivek committed:
Merge pull request #572 from durgakiran/feature/toggle-event-mirror

Add global disable/enable event mirroring

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

manojVivek committed:
Merge pull request #576 from responsively-org/all-contributors/add-durgakiran

docs: add durgakiran as a contributor

durgakiran committed:
Add screen shot all devices functionality

manojVivek commented:
@durgakiran I think, it would be more useful if we add the device name above the screenshot. What do you think? 

This will help the use case of showcasing the screenshots.

Commit:
Merge remote-tracking branch 'origin/master' into master

durgakiran commented:
> @durgakiran I think, it would be more useful if we add the device name above the screenshot. What do you think?
> 
> This will help the use case of showcasing the screenshots.

I thought about it and forgot to add that functionality. I will update the PR. any other changes you recommend?

Commit:
ZenButton z-index fixed

manojVivek committed:
Merge pull request #545 from karllabrador/531-toggle-spellcheck

Add toggle spellcheck

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

manojVivek committed:
Merge pull request #579 from responsively-org/all-contributors/add-karllabrador

docs: add karllabrador as a contributor

Commit:
Version bump 🔥 - v0.16.0

durgakiran commented:
I am converting this pull request into draft as I observed a bug.

durgakiran commented:
@manojVivek   one thing I observed was when a screenshot of a web view is happening, the screenshot of that webview is not happening until that web view is in the view port of the electron, any reason in your mind about this behaviour? It just does not seem logical to me? what do you think?

durgakiran committed:
Add Names of devices in the merged screenshot

jjavierdguezas committed:
add second shortcut for taking screenshot #575

jjavierdguezas committed:
add F5 as shortcut for reload action #571
- needed to make configurable the order of shortcut definitions
- needed to make skippable the menu shortcuts

jjavierdguezas committed:
update alternatives to download

- add winget command
- add rpm command

manojVivek committed:
Merge pull request #580 from responsively-org/feature/575

Shortcuts improvements

manojVivek commented:
> @manojVivek one thing I observed was when a screenshot of a web view is happening, the screenshot of that webview is not happening until that web view is in the view port of the electron, any reason in your mind about this behaviour? It just does not seem logical to me? what do you think?

@durgakiran Yeah that is how Chromium behaves. Maybe they don't render/paint the sections of the screen if they are outside the viewport. 
We handle this behaviour by zoom-out the web-views before the screenshot is taken so that all the web-views are in the viewport and then process the screenshot. Did you face it during your testing? Ideally, you shouldn't have.

durgakiran commented:
@manojVivek yes, yes I faced it. Maybe because I was almost completely recoded the screenshot all functionality. So, I might have missed something.

durgakiran commented:
Ok I will look into that, thanks

Commit:
Merge remote-tracking branch 'responsively/master'

Commit:
improve styles and position of url link container
use `<webview/>` event for url hover instead of custom preload code

Commit:
remove LinkHoverDisplay's style.css

durgakiran committed:
Fix screen capture when screen not in viewport

durgakiran committed:
Merge branch 'master' of https://github.com/responsively-org/responsively-app into 541-feature-screen-shot-enhancements

manojVivek commented on desktop-app/app/components/ScreenshotManager/Capture.js:
```diff
@@ -0,0 +1,268 @@
+
+  const fsUtils = new FileSystemUtils(path, defaultPath, now, address);
+
+  for (let i = 0; i < sortedWebViews.length; i += 1) {
```
@durgakiran Processing the screenshots sequentially is going to take a very long time. We should look at parallelizing the capture here.

durgakiran replied:
As far as I observed, we can't use node workers in electron. I think parallelization in Electron can only be achieved through web workers but web workers don't have access to webviews. So I think we kind of hit a big roadblock there. Do you have any suggestions?

durgakiran replied:
I could not find any alternative here

manojVivek replied:
@durgakiran Why don't you make it an `async.map` or something like that? 
So that the capture can happen in parallel whenever the CPU is free. We have some intentional `sleeps` in the capture workflow and this parallelization will cut down some waiting time over that.

manojVivek commented on desktop-app/app/components/ScreenshotManager/Capture.js:
```diff
@@ -0,0 +1,268 @@
+  images: Array<Electron.NativeImage>,
+  deviceNames: Array<string>
+) {
+  const canvas = document.createElement('canvas');
```
@durgakiran Is there a better way to do the merging instead of using a `canvas` in terms of performance? 
Just to make sure the screenshot functionality finishes as quickly as possible.

durgakiran replied:
@manojVivek I think, we should really explore here. So far I did not find any, but I believe there can be better methods. maybe we should explore a bit of web assembly.

durgakiran replied:
I am exploring `webGL`, seems promising. I will update you if I find a real good way.

durgakiran replied:
I have optimized the process a bit, the merging and saving is fast now.

manojVivek reviewed (commented):
@durgakiran Great effort in getting this to PR. 👏🏼 
 We just need to tweak the design of the merged image a bit as well so that it looks more presentable.

Maybe a `lightgray` background, `black` color for the devices' names and some margin around the final image would make it look better?

manojVivek committed:
Merge pull request #522 from rishichawda/master

feat(desktop-app/app): show url link on hover

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

manojVivek committed:
Merge pull request #581 from responsively-org/all-contributors/add-rishichawda

docs: add rishichawda as a contributor

durgakiran commented on desktop-app/app/components/ScreenshotManager/index.js:
```diff
@@ -1,5 +1,122 @@
+
+  const handleGo = () => {
+    const payload = {
+      zip,
```
I forgot to remove `zip`. I will update that

jjavierdguezas committed:
add Clear Address History button to UserSettings Menu
resolves #341

jjavierdguezas committed:
move new section to the bottom for better prev select display

jjavierdguezas committed:
Allow pass url/filepath from cli args
resolves 369

manojVivek committed:
Merge pull request #586 from responsively-org/feature/369

Allow pass url/filepath from cli args

jjavierdguezas committed:
fix typo

manojVivek committed:
Merge pull request #583 from responsively-org/feature/341

Add Clear Address History button to User Settings Menu

durgakiran committed:
Optimize image merging logic

durgakiran reviewed (commented):
As far as design changes for png background, I could not see any point in it. It unnecessarily increases the file size.

jjavierdguezas committed:
add option to swtich between blank page and homepage when app launches

jjavierdguezas committed:
Merge pull request #262 from ziabs/patch-1

Git ignore .vscode

jjavierdguezas committed:
Revert "Git ignore .vscode"

jjavierdguezas committed:
Merge pull request #591 from responsively-org/revert-262-patch-1

Revert "Git ignore .vscode"

jjavierdguezas committed:
Add Page Navigator

manojVivek committed:
Merge pull request #588 from responsively-org/feature/521

Add option to switch between blank page and homepage when app launches

jjavierdguezas committed:
Merge branch 'master' into feature/page-navigator

manojVivek committed:
Merge pull request #593 from responsively-org/feature/page-navigator

Page navigator

Commit:
Code moved from old declined PR

Commit:
Added dont ask again checkbox to the move to applications message box

Commit:
Restricted move to applications functionality to MacOS only

manojVivek committed:
Merge pull request #595 from responsively-org/feature/macos-move-to-applications

Feature/macos move to applications

jjavierdguezas committed:
fix notification styles

manojVivek committed:
Merge pull request #596 from responsively-org/fix/584

fix notification styles

durgakiran committed:
Merge branch 'master' of https://github.com/responsively-org/responsively-app into 541-feature-screen-shot-enhancements

manojVivek commented:
> As far as design changes for png background, I could not see any point in it. It unnecessarily increases the file size.

We can save the image as `jpeg` and that will have reduced file size. 
The reason why I'm insisting it is having non-transparent background is that is going to make it easier for the users to just share/use it without any additional processing.

durgakiran committed:
Update screen loading logic

rille111 commented:
Hey still is in PR.. would appreciate this functionality. :)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/responsively-org/responsively-app?pullRequest=578) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/responsively-org/responsively-app?pullRequest=578) before we can accept your contribution.<br/>**2** out of **8** committers have signed the CLA.<br/><br/>:white_check_mark: durgakiran<br/>:white_check_mark: manojVivek<br/>:x: karllabrador<br/>:x: ziabs<br/>:x: rishichawda<br/>:x: Manoj Vivek<br/>:x: jjavierdguezas<br/>:x: allcontributors[bot]<br/><hr/>**Manoj Vivek** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/responsively-org/responsively-app?pullRequest=578) it.</sub>

durgakiran commented:
@manojVivek  thought of picking this up again but found a lot has been changed, since this PR. Will work on this PR again

manojVivek commented:
@durgakiran Great to see you back again. 
Yeah the app has went through a full revamp, and should be relatively easier to add new features.
